--- a/arch/arm/common/gic_old.c	2012-12-20 13:18:49.167490811 +0530
+++ b/arch/arm/common/gic.c	2012-12-20 13:19:29.851491456 +0530
@@ -158,6 +158,11 @@ static void gic_unmask_irq(struct irq_da
 {
 	u32 mask = 1 << (gic_irq(d) % 32);
 
+   if(gic_irq(d) == 82) {
+       printk("gic unmask irq; do nothing for uart\n");
+       return;
+   }
+
 	raw_spin_lock(&irq_controller_lock);
 	if (gic_arch_extn.irq_unmask)
 		gic_arch_extn.irq_unmask(d);
